command/snapshot/inspect/snapshot_inspect.go

```diff
@@ -56,27 +65,181 @@ func (c *cmd) Run(args []string) int {
 	}
 	defer f.Close()
 
-	meta, err := snapshot.Verify(f)
+	readFile, meta, err := snapshot.Read(nil, f)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error reading snapshot: %s", err))
+	}
+	stats, offset, err := enhance(readFile)
 	if err != nil {
-		c.UI.Error(fmt.Sprintf("Error verifying snapshot: %s", err))
+		c.UI.Error(fmt.Sprintf("Error extracting snapshot data: %s", err))
 		return 1
 	}
+	// Outputs the original style of inspect information
+	legacy, err := c.legacyStats(meta)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error outputting snapshot data: %s", err))
+	}
+	c.UI.Info(legacy.String())
+
+	// Outputs the more detailed snapshot information
+	enhanced, err := c.readStats(stats, offset)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error reading snapshot stats: %s", err))
+		return 1
+	}
+	c.UI.Info(enhanced.String())
+
+	return 0
+}
 
+// legacyStats outputs the expected stats from the original snapshot
+// inspect command
+func (c *cmd) legacyStats(meta *raft.SnapshotMeta) (bytes.Buffer, error) {
 	var b bytes.Buffer
 	tw := tabwriter.NewWriter(&b, 0, 2, 6, ' ', 0)
 	fmt.Fprintf(tw, "ID\t%s\n", meta.ID)
 	fmt.Fprintf(tw, "Size\t%d\n", meta.Size)
 	fmt.Fprintf(tw, "Index\t%d\n", meta.Index)
 	fmt.Fprintf(tw, "Term\t%d\n", meta.Term)
 	fmt.Fprintf(tw, "Version\t%d\n", meta.Version)
-	if err = tw.Flush(); err != nil {
+	if err := tw.Flush(); err != nil {
 		c.UI.Error(fmt.Sprintf("Error rendering snapshot info: %s", err))
-		return 1
+		return b, err
 	}
+	return b, nil
+}
+
+type typeStats struct {
+	Name  string
+	Sum   int
+	Count int
+}
 
-	c.UI.Info(b.String())
+// countingReader helps keep track of the bytes we have read
+// when reading snapshots
+type countingReader struct {
+	wrappedReader io.Reader
+	read          int
+}
 
-	return 0
+func (r *countingReader) Read(p []byte) (n int, err error) {
+	n, err = r.wrappedReader.Read(p)
+	if err == nil {
+		r.read += n
+	}
+	return n, err
+}
+
+// enhance utilizes ReadSnapshot to populate the struct with
+// all of the snapshot's itemized data
+func enhance(file io.Reader) (map[structs.MessageType]typeStats, int, error) {
+	stats := make(map[structs.MessageType]typeStats)
+	var offset int
+	cr := &countingReader{wrappedReader: file}
+	offset = 0
```

Comment:
minor: this line does nothing because `var offset int` already sets the value to 0, this line can be removed.